libraries/chain/fork_database.cpp

```diff
@@ -436,13 +494,18 @@ namespace eosio { namespace chain {
          result.second.push_back(second_branch);
       }
       return result;
-   } /// fetch_branch_from
+   } /// fetch_branch_from_impl
 
    /// remove all of the invalid forks built off of this id including this id
    void fork_database::remove( const block_id_type& id ) {
+      std::lock_guard g( my->mtx );
+      return my->remove_impl( id );
+   }
+
+   void fork_database_impl::remove_impl( const block_id_type& id ) {
       deque<block_id_type> remove_queue{id};
-      const auto& previdx = my->index.get<by_prev>();
-      const auto& head_id = my->head->id;
+      const auto& previdx = index.get<by_prev>();
```

Comment:
for the `remove_queue`, is there a benefit to use `std::deque` vs `std::vector`?

Reply:
Yes, this was a performance improvement awhile back. Significant performance improvement using `deque` over `vector` for large number of transactions. Note it is `deque` not `std::deque`, we actually use a `boost` `deque` for modern `boost` versions...
Or at least we are suppose to be. I guess that never got ported over to leap.
https://github.com/EOSIO/eos/blob/master/libraries/chain/include/eosio/chain/types.hpp#L85
I created https://github.com/AntelopeIO/leap/issues/576 for this.